source/config/__init__.py

```diff
@@ -167,6 +168,19 @@ def initConfigPath(configPath=None):
 		configPath=globalVars.appArgs.configPath
 	if not os.path.isdir(configPath):
 		os.makedirs(configPath)
+	else:
+		OLD_CODE_DIRS = ("appModules", "brailleDisplayDrivers", "globalPlugins", "synthDrivers")
+		# #10014: Since #9238 code from these directories is no longer loaded.
+		# However they still exist in config for older installations. Remove them if empty to minimize confusion.
+		for dir in OLD_CODE_DIRS:
+			dir = os.path.join(configPath, dir)
+			if os.path.isdir(dir):
+				try:
+					os.rmdir(dir)
+					log.info("Removed old plugins dir: %s", dir)
+				except OSError as ex:
+					if ex.errno == errno.ENOTEMPTY:
+						log.info("Failed to remove old plugins dir: %s. Directory not empty.", dir)

```

Comment:
For people with files in these folders, this will be done every startup, slowing down the startup of NVDA. I'll have to think about this a little. Perhaps you can profile this to see how long it takes?

Reply:
@feerrenrut wrote:
> Perhaps you can profile this to see how long it takes?

I've timed this function in two scenarios:
- Best case - no old  folders
- worst case - files in all four of them


In the best case execution took (on average for 5 launches) 0.0002 sec
for the worst case (the same sample) 0.001 sec



On paper the time increase is big, but I do not think it would be very noticeable. For most people these folders are empty either because they never used them, or because they moved files from them to the scratchpad.

Reply:
While the impact of this is unlikely to be noticeable, I think it would be good to clean up these files in all cases since they are not used at all. An alternative approach would be to do this in the config upgrade step. You'd have to bump the config version number. This way it will only be done once per user profile. Rather than leaving the files in place they could be moved to an "Archived-unused" directory and a warning for the user?

Actually, the config upgrade step and version bump is not necessary, though it's probably cleaner, since it maintains a history of changes in the user config.

Reply:
@feerrenrut wrote:

> An alternative approach would be to do this in the config upgrade step. You'd have to bump the config version number. This way it will only be done once per user profile. 

I was thinking about something along these lines, however to address all cases it might be worth to do it as part of NVDA install, this has an added  advantage of cleaning systemConfig as well, and also as a config upgrade step.

> Rather than leaving the files in place they could be moved to an "Archived-unused" directory and a warning for the user?

I don't think this is a good idea. There are essentially 3 states in which config can be:
1. These folders are not present,, because config has been created after NVDA 2019.1
2. These folders are present, but they are empty, either because they were never used, or because files from them were moved to the scratchpad.
3. These folders are not empty, because someone uses his config with old and new NVDA versions interchangeably - expecting users to know how to convert they plugins to addons is not reasonable.

Moving these old plugins to the archive would break the third use-case, and even though it is not a very common one keeping config backwards compatible is important in my opinion.

Reply:
> I was thinking about something along these lines, however to address all cases it might be worth to do it as part of NVDA install, this has an added advantage of cleaning systemConfig as well, and also as a config upgrade step.

I'm pretty sure that doing it in the config upgrade will also affect the system profile the first time the system instance of NVDA starts.

> These folders are not empty, because someone uses his config with old and new NVDA versions interchangeably - expecting users to know how to convert they plugins to addons is not reasonable.

This is really not something we support. I understand that some users do this, and there aren't normally many issues to do it, but it should be a decision in its own right to support this use case. On the other hand, its not really that important to clean up these directories. Perhaps for now they should just be deleted if empty, and sometime in the future (eg 2022.1?) non-empty folders should be archived.